wasm/section.go

```diff
@@ -0,0 +1,696 @@
+			m.Memory.Section = s
+		}
+	case SectionIDGlobal:
+		logger.Println("section global")
+		if err = m.readSectionGlobals(sectionReader); err == nil {
+			m.Global.Section = s
+		}
+	case SectionIDExport:
+		logger.Println("section export")
+		if err = m.readSectionExports(sectionReader); err == nil {
+			m.Export.Section = s
+		}
+	case SectionIDStart:
+		logger.Println("section start")
+		if err = m.readSectionStart(sectionReader); err == nil {
+			m.Start.Section = s
+		}
+	case SectionIDElement:
+		logger.Println("section element")
+		if err = m.readSectionElements(sectionReader); err == nil {
+			m.Elements.Section = s
+		}
+	case SectionIDCode:
+		logger.Println("section code")
+		if err = m.readSectionCode(sectionReader); err == nil {
+			m.Code.Section = s
+		}
+	case SectionIDData:
+		logger.Println("section data")
+		if err = m.readSectionData(sectionReader); err == nil {
+			m.Data.Section = s
+		}
+	default:
+		return false, ErrInvalidSectionID(s.ID)
+	}
+
+	return false, err
+}
+
+// SectionTypes declares all function signatures that will be used in a module.
+type SectionTypes struct {
+	Section
+	Entries []FunctionSig
+}
+
+func (m *Module) readSectionTypes(r io.Reader) error {
+	s := &SectionTypes{}
+	count, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return err
+	}
+
+	s.Entries = make([]FunctionSig, int(count))
+
+	for i := uint32(0); i < count; i++ {
+		if s.Entries[i], err = readFunction(r); err != nil {
+			return err
+		}
+
+	}
+
+	m.Types = s
+
+	return nil
+}
+
+// SectionImports declares all imports that will be used in the module.
+type SectionImports struct {
+	Section
+	Entries []ImportEntry
+}
+
+func (m *Module) readSectionImports(r io.Reader) error {
+	s := &SectionImports{}
+	count, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return err
+	}
+	s.Entries = make([]ImportEntry, count)
+
+	for i := uint32(0); i < count; i++ {
+		var err error
+		s.Entries[i], err = readImportEntry(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	m.Import = s
+	return nil
+}
+
+func readImportEntry(r io.Reader) (ImportEntry, error) {
+	i := ImportEntry{}
+
+	modLen, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return i, err
+	}
+
+	if i.ModuleStr, err = readString(r, uint(modLen)); err != nil {
+		return i, err
+	}
+
+	fieldLen, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return i, err
+	}
+
+	if i.FieldStr, err = readString(r, uint(fieldLen)); err != nil {
+		return i, err
+	}
+
+	if i.Kind, err = readExternal(r); err != nil {
+		return i, err
+	}
+
+	var ptr *GlobalVar
+
+	switch i.Kind {
+	case ExternalFunction:
+		logger.Println("importing function")
+		i.Type, err = leb128.ReadVarUint32(r)
+	case ExternalTable:
+		logger.Println("importing table")
+		i.Type, err = readTable(r)
+	case ExternalMemory:
+		logger.Println("importing memory")
+		i.Type, err = readMemory(r)
+	case ExternalGlobal:
+		logger.Println("importing global var")
+		ptr, err = readGlobalVar(r)
+		if ptr != nil {
```

Comment:
From `readGlobalVar`'s code, that shouldn't happen. Anyway, I've added a better check for `err != nil` instead.